sycl/test-e2e/Matrix/joint_matrix_prefetch_impl.hpp

```diff
@@ -0,0 +1,141 @@
+    joint_matrix_load(sg, sub_b,
+                      pB + k * N + sg_starty / SG_SZ * TN * vnniFactor,
+                      N * vnniFactor);
+    joint_matrix_mad(sg, sub_c, sub_a, sub_b, sub_c);
+  }
+  joint_matrix_store(sg, sub_c,
+                     pC + (sg_startx * TM) * N + sg_starty / SG_SZ * TN, N,
+                     layout::row_major);
+}
+
+template <typename T1, typename T2, typename T, size_t M, size_t N, size_t K>
+void joint_matrix_gemm(sub_group sg, size_t sg_startx, size_t sg_starty, T1 *A,
+                       T2 *B, T *C) {
+  auto pA = address_space_cast<sycl::access::address_space::global_space,
+                               access::decorated::no>(A);
+  auto pB = address_space_cast<sycl::access::address_space::global_space,
+                               access::decorated::no>(B);
+  auto pC = address_space_cast<sycl::access::address_space::global_space,
+                               access::decorated::no>(C);
+
+  joint_matrix<sub_group, T1, use::a, TM, TK, layout::row_major> sub_a;
+  joint_matrix<sub_group, T2, use::b, TK, TN, layout::row_major> sub_b;
+  joint_matrix<sub_group, T, use::accumulator, TM, TN> sub_c;
+  joint_matrix_prefetch(sg, A + (sg_startx * TM) * K, K, layout::row_major, TM,
+                        TK, syclex::properties{syclex::prefetch_hint_L1});
+  joint_matrix_prefetch(sg, B + sg_starty / SG_SZ * TN, N, layout::row_major,
+                        TK, TN, syclex::properties{syclex::prefetch_hint_L2});
+  joint_matrix_prefetch(sg, C + (sg_startx * TM) * N + sg_starty / SG_SZ * TN,
+                        N, layout::row_major, TM, TN,
+                        syclex::properties{syclex::prefetch_hint_L1});
+  joint_matrix_fill(sg, sub_c, 1);
+  for (int k = 0; k < K; k += TK) {
+    joint_matrix_load(sg, sub_a, pA + (sg_startx * TM) * K + k, K);
+    joint_matrix_load(sg, sub_b, pB + k * N + sg_starty / SG_SZ * TN, N);
+    joint_matrix_mad(sg, sub_c, sub_a, sub_b, sub_c);
+  }
+  joint_matrix_store(sg, sub_c,
+                     pC + (sg_startx * TM) * N + sg_starty / SG_SZ * TN, N,
+                     layout::row_major);
+}
+
+template <typename T, typename T1, typename T2, size_t M, size_t N, size_t K,
+          unsigned int vnniFactor>
+void matrix_multiply(T *C, T1 *A, T2 *B, queue q) {
+  size_t NDRangeM = M / TM;
+  size_t NDRangeN = N / TN;
+
+  q.submit([&](handler &cgh) {
+     cgh.parallel_for(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [=](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           if constexpr (vnniFactor > 1)
```

Comment:
if we unify joint_matrix_gemm_vnni as I suggested above and also propagate necessary parameters to matrix_multiply signature, we don't need this condition at all, we can just use one call, something like
joint_matrix_gemm_vnni<T1, T2, T, M, N, K, vnniFactor, B_layout, maybe_prefetch_hint>(sg, sg_startx, sg_starty, A, B, C);